src/MemAllocHeap.cpp

```diff
@@ -1,37 +1,146 @@
 #include "MemAllocHeap.hpp"
 #include <stdlib.h>
+#include <stdio.h>
 
+#if defined(TUNDRA_WIN32)
+#include "DbgHelp.h"
+#endif
 
+#define LOG_ALLOC 0
+
+#if LOG_ALLOC && TUNDRA_WIN32
+
+static bool GetStackWalk( std::string &outWalk )
+{
+    // Set up the symbol options so that we can gather information from the current
+    // executable's PDB files, as well as the Microsoft symbol servers.  We also want
+    // to undecorate the symbol names we're returned.  If you want, you can add other
+    // symbol servers or paths via a semi-colon separated list in SymInitialized.
+    ::SymSetOptions( SYMOPT_DEFERRED_LOADS | SYMOPT_INCLUDE_32BIT_MODULES | SYMOPT_UNDNAME );
+    if (!::SymInitialize( ::GetCurrentProcess(), "http://msdl.microsoft.com/download/symbols", TRUE )) return false;
+
+    // Capture up to 25 stack frames from the current call stack.  We're going to
+    // skip the first stack frame returned because that's the GetStackWalk function
+    // itself, which we don't care about.
+    PVOID addrs[ 25 ] = { 0 };
+    USHORT frames = CaptureStackBackTrace( 1, 25, addrs, NULL );
+
+    for (USHORT i = 0; i < frames; i++) {
+        // Allocate a buffer large enough to hold the symbol information on the stack and get
+        // a pointer to the buffer.  We also have to set the size of the symbol structure itself
+        // and the number of bytes reserved for the name.
+        ULONG64 buffer[ (sizeof( SYMBOL_INFO ) + 1024 + sizeof( ULONG64 ) - 1) / sizeof( ULONG64 ) ] = { 0 };
+        SYMBOL_INFO *info = (SYMBOL_INFO *)buffer;
+        info->SizeOfStruct = sizeof( SYMBOL_INFO );
+        info->MaxNameLen = 1024;
+
+        // Attempt to get information about the symbol and add it to our output parameter.
+        DWORD64 displacement = 0;
+        if (::SymFromAddr( ::GetCurrentProcess(), (DWORD64)addrs[ i ], &displacement, info )) {
+            outWalk.append( info->Name, info->NameLen );
+            outWalk.append( "\n" );
+        }
+    }
+
+    ::SymCleanup( ::GetCurrentProcess() );
+
+    return true;
+}
+
+static void print_trace()
+{
+    std::string output;
+    GetStackWalk(output);
+    printf("trace: %s\n", output.c_str());
+}
+
+#endif
+
+static uint64_t s_ActiveHeaps = 0;
+
+void HeapVerifyNoLeaks()
+{
+    if (s_ActiveHeaps != 0)
+        Croak("%d heaps have been initialized but not destroyed.", s_ActiveHeaps);
+}
 
 void HeapInit(MemAllocHeap *heap)
 {
-    (void)heap;
+    AtomicAdd(&s_ActiveHeaps, 1);
+#if DEBUG_HEAP
+    heap->m_Size = 0;
+#endif
 }
 
 void HeapDestroy(MemAllocHeap *heap)
 {
+    AtomicAdd(&s_ActiveHeaps, -1);
+#if DEBUG_HEAP
+    if (heap->m_Size != 0)
+    {
+        if (getenv("BEE_ENABLE_TUNDRA_HEAP_VALIDATION"))
```

Comment:
nitpick: name it `TUNDRA_HEAP_VALIDATION` ?

Reply:
I named it BEE_ for consistency with other environment variables. Lucas says that our branched-off tundra really only exists in the context of Bee (and some of these are set by the process invoking bee, which shouldn't need to know about TUNDRA underneath). I don't care much except that it would be nice if we get them to be somewhat consistent.

Reply:
yeah, in terms of env vars, we can consider this fork of tundra part of bee, that should move to the same repo as the bee code as soon as bee moves out of the unity repo